### PR TITLE
feat: Add address reputation

### PR DESCRIPTION
**Pull request by alex-stone**
### What changed? Why?
This adds support for fetching an address' reputation lazily.

Example:
```ruby
risky_address = Coinbase::ExternalAddress.new(:ethereum_mainnet, '0x12846c6Fd6baBFE4bC6F761eB871eFfFDEb26913')

# Returns the reputation of the address
risky_address.reputation
=> Coinbase::AddressReputation{score: '-90', total_transactions: '0', unique_days_active: '0', longest_active_streak: '0', current_active_streak: '0', activity_period_days: '0', token_swaps_performed: '0', bridge_transactions_performed: '0', lend_borrow_stake_transactions: '0', ens_contract_interactions: '0', smart_contract_deployments: '0'}

# Returns whether the address itself is deemed "risky"
risky_address.risky?
=> true
```

### CODE REVIEW

**Comment by cb-heimdall** [comment]
<h1>🟡 Heimdall Review Status</h1>
<table>
<tr>
<th>Requirement</th>
<th>Status</th>
<th>More Info</th>
</tr>
<tr>
<td>Reviews</td>
<td>
🟡
<code>0/1</code>
</td>
<td>
<details>
<summary>Denominator calculation</summary>
<table>
<tr>
<td>
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><code>1</code> if user is bot</td>
<td>0</td>
</tr>
<tr>
<td><code>1</code> if user is external</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>.codeflow.yml</code>
</td>
<td>1</td>
</tr>
<tr>
<td>
Additional review requirements
<details>
<summary>Show calculation</summary>
<table>
<tr>
<td><strong>Max</strong></td>
<td><strong>0</strong></td>
</tr>
</table>
</details>
</td>
<td>0</td>
</tr>
<tr>
<td>
From <code>CODEOWNERS</code>
</td>
<td>0</td>
</tr>
<tr>
<td>Global minimum</td>
<td>0</td>
</tr>
<tr>
<td><strong>Max</strong></td>
<td>
<strong>
1
</strong>
</td>
</tr>
</table>
</details>
</td>
<td>1</td>
</tr>
<tr>
<td><code>1</code> if commit is unverified</td>
<td>0</td>
</tr>
<tr>
<td><strong>Sum</strong></td>
<td><strong>1</strong></td>
</tr>
</table>
</details>
</td>
</tr>
</table>

File: spec/unit/coinbase/address_reputation_spec.rb [hunk 1]
```diff
@@ -0,0 +1,94 @@
+# frozen_string_literal: true
+
+describe Coinbase::AddressReputation do
+  let(:address_id) { '0x123456789abcdef' }
+  let(:network) { 'ethereum-mainnet' }
+  let(:mock_model) do
+    instance_double(
+      Coinbase::Client::AddressReputation,
```
Comment (alex-stone) [inline spec/unit/coinbase/address_reputation_spec.rb:8]: We should add a factory for `AddressReputation`: https://github.com/coinbase/coinbase-sdk-ruby/blob/master/spec/factories/fund_quote.rb

File: spec/unit/coinbase/address_reputation_spec.rb [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+      reputation = described_class.new(mock_model)
+      metadata = reputation.metadata
+
+      expect(metadata.to_h[:total_transactions]).to eq(1)
+    end
+  end
+
+  describe '#risky?' do
+    it 'returns false for a positive score' do
+      reputation = described_class.new(mock_model)
```
Comment (alex-stone) [inline spec/unit/coinbase/address_reputation_spec.rb:76]: Please use the `subject` test helper for this situation

https://github.com/coinbase/coinbase-sdk-ruby/blob/6b92ca0cb3c9c590e816196e96fc56c10a907dff/spec/unit/coinbase/fiat_amount_spec.rb#L38

File: spec/unit/coinbase/address_reputation_spec.rb [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+      expect(reputation.risky?).to be false
+    end
+
+    it 'returns true for a negative score' do
+      allow(mock_model).to receive(:score).and_return(-10)
```
Comment (alex-stone) [inline spec/unit/coinbase/address_reputation_spec.rb:81]: This sort of setup should be done using the factory, e.g.

```
describe '#risky?' do
  context 'when the score is positive' do
    let(:mock_model) { build(:address_reputation, score: 42) {

    it 'returns false' do
      expect(reputation).not_to be_risky
    end
  end  

  context 'when the score is negative' 
    let(:mock_model) { build(:address_reputation, score: -10) }
    
    it 'returns true' do
      expect(reputation).to be_risky
    end
  end
end
```

File: spec/unit/coinbase/address_reputation_spec.rb [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      metadata = reputation.metadata
+
+      expect(metadata).to be_a(Coinbase::AddressReputation::Metadata)
+    end
+
+    it 'has correct metadata values for total transactions' do
+      reputation = described_class.new(mock_model)
+      metadata = reputation.metadata
+
+      expect(metadata.to_h[:total_transactions]).to eq(1)
```
Comment (alex-stone) [inline spec/unit/coinbase/address_reputation_spec.rb:70]:
```
expect(reputation.metadata.total_transactions).to be(1)
```

**Comment by alex-stone** [comment]
We also will want to add an address reputation shared example:
https://github.com/coinbase/coinbase-sdk-ruby/blob/master/spec/support/shared_examples/address_balances.rb

And add them to the following tests:
https://github.com/coinbase/coinbase-sdk-ruby/blob/6b92ca0cb3c9c590e816196e96fc56c10a907dff/spec/unit/coinbase/address_spec.rb#L45-L47
https://github.com/coinbase/coinbase-sdk-ruby/blob/6b92ca0cb3c9c590e816196e96fc56c10a907dff/spec/unit/coinbase/address/wallet_address_spec.rb#L101-L103
https://github.com/coinbase/coinbase-sdk-ruby/blob/6b92ca0cb3c9c590e816196e96fc56c10a907dff/spec/unit/coinbase/address_spec.rb#L45-L47

**Comment by arpitsrivastava-cb** [comment]
> We also will want to add an address reputation shared example: https://github.com/coinbase/coinbase-sdk-ruby/blob/master/spec/support/shared_examples/address_balances.rb
> 
> And add them to the following tests:
> 
> https://github.com/coinbase/coinbase-sdk-ruby/blob/6b92ca0cb3c9c590e816196e96fc56c10a907dff/spec/unit/coinbase/address_spec.rb#L45-L47
> 
> 
> https://github.com/coinbase/coinbase-sdk-ruby/blob/6b92ca0cb3c9c590e816196e96fc56c10a907dff/spec/unit/coinbase/address/wallet_address_spec.rb#L101-L103
> 
> 
> https://github.com/coinbase/coinbase-sdk-ruby/blob/6b92ca0cb3c9c590e816196e96fc56c10a907dff/spec/unit/coinbase/address_spec.rb#L45-L47

Done

File: spec/support/shared_examples/address_reputation.rb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+  it 'fetches the address reputation from the API' do
+    expect(address.reputation).to be_a(Coinbase::AddressReputation)
+  end
+
+  it 'returns the reputation score' do
+    expect(address_reputation.score).to eq(score)
+  end
+
+  it 'returns metadata as a Metadata object' do
```
Comment (jazz-cb) [inline spec/support/shared_examples/address_reputation.rb:20]: should we add a test for `risky?`

**Review by jazz-cb — DISMISSED** [review]

**Review by jazz-cb — APPROVED** [review]

**Pull request merged**